### ISSUE TITLE
Chainstate: node crash due to Sqlite database lock

### ISSUE THREAD

**Issue by jcnelson** [blockstack/stacks-blockchain#1939]
Somewhere we're not setting a busy handler in chainstate processing:

```
DEBUG [1601641208.564] [src/chainstate/stacks/db/transactions.rs:849] [ThreadId(6)] Process transaction 6a5a60313a37e2970e8f512cd9191c9e972ca54b452e4d88dd0166c5d7c9a554 (Coinbase)
ERROR [1601641208.568] [src/vm/database/sqlite.rs:30] [ThreadId(6)] Failed to insert/replace (f94747b71a877b0d2f8e1e9b437c5ca012db4d242ec8921d9a39c7306d5b9a0c0000000000000000,0000000000000000000002bbc5c28700000000000000000000000000000000000000000000000000): SqliteFailure(Error { code: DatabaseBusy, extended_code: 5 },
 Some("database is locked"))                                                                                                                                   
Process abort due to thread panic
   0: stacks_node::main::{{closure}}                                                                                                                                                                                                                                                                                           
             at src/main.rs:51                                                                                                                                                                                                                                                                                                 
   1: std::panicking::rust_panic_with_hook                                                                                                                                                                                                                                                                                     
   2: std::panicking::begin_panic                                                                                                                                                                                                                                                                                              
             at /home/buildozer/aports/community/rust/src/rustc-1.44.0-src/src/libstd/panicking.rs:438                                                                                                                                                                                                                         
   3: blockstack_lib::vm::database::sqlite::sqlite_put                                                                                                                                                                                                                                                                         
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/vm/database/sqlite.rs:31                                                                                                                                                                                                                           
   4: blockstack_lib::vm::database::sqlite::SqliteConnection::put                                                                                                                                                                                                                                                              
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/vm/database/sqlite.rs:62                                                                                                                                                                                                                           
   5: <blockstack_lib::vm::database::marf::MarfedKV as blockstack_lib::vm::database::marf::ClarityBackingStore>::put_all                                                                                                                                                                                                       
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/vm/database/marf.rs:509                                                                                                                                                                                                                            
   6: blockstack_lib::vm::database::key_value_wrapper::RollbackWrapper::commit                                                                                 
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/vm/database/key_value_wrapper.rs:230                                                                                                                                                                                                               
   7: blockstack_lib::vm::clarity::ClarityTransactionConnection::commit                                                                                                                                                                                                                                                        
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/vm/clarity.rs:835                                                                                                                                                                                                                                  
   8: blockstack_lib::chainstate::stacks::db::transactions::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_transaction                                                                                                                                                                               
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/chainstate/stacks/db/transactions.rs:885                                                                                                                                                                                                           
   9: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_block_transactions                                                                                                                                                                              
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/chainstate/stacks/db/blocks.rs:3682                                                                                                                                                                                                                
  10: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::append_block                                                                                                                                                                                            
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/chainstate/stacks/db/blocks.rs:3886                                                                                                                                                                                                                
  11: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_next_staging_block                                                                                                                                                                              
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/chainstate/stacks/db/blocks.rs:4306                                                                                                                                                                                                                
  12: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_blocks                                                                                                                                                                                          
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/chainstate/stacks/db/blocks.rs:4464                                                                                                                                                                                                                
  13: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,N,U>::process_ready_blocks                                                                                                                                                                                                                                  
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/chainstate/coordinator/mod.rs:559                                                                                                                                                                                                                  
  14: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,N,U>::handle_new_stacks_block                                                                                                                                                                                                                               
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/chainstate/coordinator/mod.rs:361                                                                                                                                                                                                                  
  15: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,blockstack_lib::chainstate::coordinator::comm::ArcCounterCoordinatorNotices,blockstack_lib::chainstate::coordinator::OnChainRewardSetProvider>::run                                                                                                         
             at /home/jude/Desktop/research/git/blockstack-ng.feat-1754/src/chainstate/coordinator/mod.rs:231                                                                                                                                                                                                                  
  16: stacks_node::run_loop::neon::RunLoop::start::{{closure}}                                                                                                                                                                                                                                                                 
             at src/run_loop/neon.rs:578                                                                                                                                                                                                                                                                                       
  17: std::sys_common::backtrace::__rust_begin_short_backtrace                                                                                                                                                                                                                                                                 
             at /home/buildozer/aports/community/rust/src/rustc-1.44.0-src/src/libstd/sys_common/backtrace.rs:130                                                                                                                                                                                                              
  18: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}                                                                                                                                                                                                                                                          
             at /home/buildozer/aports/community/rust/src/rustc-1.44.0-src/src/libstd/thread/mod.rs:475                                                                                                                                                                                                                        
  19: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once                                                                                                                                                                                                                                          
             at /home/buildozer/aports/community/rust/src/rustc-1.44.0-src/src/libstd/panic.rs:318                                                                                                                                                                                                                             
  20: std::panicking::try::do_call                                                                                                                                                                                                                                                                                             
             at /home/buildozer/aports/community/rust/src/rustc-1.44.0-src/src/libstd/panicking.rs:331                                                                                                                                                                                                                         
  21: __rust_try                                                                                                                                                                                                                                                                                                               
  22: std::panicking::try                                                                                                                                                                                                                                                                                                      
             at /home/buildozer/aports/community/rust/src/rustc-1.44.0-src/src/libstd/panicking.rs:274                                                                                                                                                                                                                         
  23: std::panic::catch_unwind                                                                                                                                                                                                                                                                                                 
             at /home/buildozer/aports/community/rust/src/rustc-1.44.0-src/src/libstd/panic.rs:394                                                                                                                                                                                                                             
  24: std::thread::Builder::spawn_unchecked::{{closure}}                                                                                                                                                                                                                                                                       
             at /home/buildozer/aports/community/rust/src/rustc-1.44.0-src/src/libstd/thread/mod.rs:474                                                                                                                                                                                                                        
  25: core::ops::function::FnOnce::call_once{{vtable.shim}}                                                                                                                                                                                                                                                                    
             at /home/buildozer/aports/community/rust/src/rustc-1.44.0-src/src/libcore/ops/function.rs:232                                                                                                                                                                                                                     
  26: std::sys::unix::thread::Thread::new::thread_start                                                                                                                                                                                                                                                                        
  27: <unknown>
```

**Comment by jcnelson** [comment]
Capturing conversation on this from Discord:

> [12:45 PM] jude: also, it seems this can happen even if we set busy handlers in all sqlite connections:
`
The presence of a busy handler does not guarantee that it will be invoked when there is lock contention. If SQLite determines that invoking the busy handler could result in a deadlock, it will go ahead and return SQLITE_BUSY to the application instead of invoking the busy handler. Consider a scenario where one process is holding a read lock that it is trying to promote to a reserved lock and a second process is holding a reserved lock that it is trying to promote to an exclusive lock. The first process cannot proceed because it is blocked by the second and the second process cannot proceed because it is blocked by the first. If both processes invoke the busy handlers, neither will make any progress. Therefore, SQLite returns SQLITE_BUSY for the first process, hoping that this will induce the first process to release its read lock and allow the second process to proceed.
`

> per: http://www.sqlite.org/c3ref/busy_handler.html

> [12:46 PM] jude: I have a suspicion: our use of SAVEPOINT and RELEASE SAVEPOINT for the Clarity VM may be leading to a deadlock if SAVEPOINT does not immediately try to acquire an exclusive lock on the Clarity VM's sidestore(edited)

> [1:00 PM] jude: and it looks like we're doing SAVEPOINTs wrong.  I think we need to be doing them like this:
> `BEGIN IMMEDIATE; -- exclusively lock the DB eagerly!`
> `SAVEPOINT xxx;`
> `-- process a transaction and do clarity stuff`
> `RELEASE SAVEPOINT xxx;`
> `SAVEPOINT yyy;`
> `-- process the next transaction and do clarity stuff`
> `RELEASE SAVEPOINT yyy;`
> `COMMIT;`

> Per the sqlite forum: http://sqlite.1065341.n5.nabble.com/SAVEPOINT-savepoint-name-DEFERRED-IMMEDIATE-EXCLUSIVE-td95302.html

> [1:02 PM] jude: (N.B. in most databases, SAVEPOINT does not start a transaction. This is a sqlite-ism. In sqlite, there is no such thing as SAVEPOINT xxx IMMEDIATE -- we'll need to guard the sequence of savepoints with a BEGIN IMMEDIATE to force an eager lock acquisition, which is necessary in order to make busy handlers behave correctly)

**Comment by jcnelson** [comment]
The reason for this crash is fundamentally due to the fact that the side-store and Clarity MARF are updated through separate transactions, whereas they should be handled through the same transaction (like how we now do with the sortition DB and headers DB).

**Comment by gaviriar** [comment]
I have observed the same issue while running krypton testnet node.

System Info:
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.1 LTS
Release:	18.04
Codename:	bionic

stacks-blockchain version:
commit: `a4deb7ac8` (tag: v23.0.0.10-krypton)

```
INFO [1603837593.366653] [testnet/stacks-node/src/neon_node.rs:1117] [ThreadId(1)] Received burnchain block #1265 including block_commit_op - n2XR5USpowuFtpFReG9pSM6ChHWHPszeAJ (2c292bbbe404037e8093622d2a26e6eec2f059532708c9c2d9645737bc659601)
INFO [1603837593.366690] [testnet/stacks-node/src/neon_node.rs:1117] [ThreadId(1)] Received burnchain block #1265 including block_commit_op - n4MVvptXe4Yiis2YQWjJaeBFNHvod13xTY (856a5dd13af812da0724a7ff4d239cd39742acee0eae98dbef15947bd42a660f)
INFO [1603837593.366713] [testnet/stacks-node/src/neon_node.rs:1117] [ThreadId(1)] Received burnchain block #1265 including block_commit_op - mrfTqmiywagAyHTgqdgaqDoqQ84pFp9BTb (5ef42bdddc4e9914fd2bb80cb4529ea5b7d8b3636b5c59ef2f9055d3ceb1e3b7)
INFO [1603837593.366733] [testnet/stacks-node/src/neon_node.rs:1103] [ThreadId(1)] Received burnchain block #1265 including block_commit_op (winning) - mohCozk3Q7gJWPA6R7gXyckywzxBez6wyL (19553bc8acb748c84b61f0b65757e7b22d7814ea984465598a194263c08ba2ac)
INFO [1603837593.366770] [testnet/stacks-node/src/neon_node.rs:1117] [ThreadId(1)] Received burnchain block #1265 including block_commit_op - mmhDWEUU13Xeb1vSqrnScx5YwPcHYb6MUK (98c1dcb960dd72d925207f269cc19f707ebf64358f121052da937b344b958ce4)
INFO [1603837593.367025] [testnet/stacks-node/src/neon_node.rs:753] [ThreadId(1)] Sleeping 5000 before issuing tenure
INFO [1603837598.353774] [src/net/relay.rs:997] [ThreadId(7)] Processing newly received blocks: 3
ERRO [1603837598.469133] [src/vm/database/sqlite.rs:46] [ThreadId(6)] Failed to insert/replace (bdd5e88ba6a9a1f99177ed01825a0a037a963a6a7c7f7eafacd20dfa5ca6b9e00000000000000000,0000000000000000000002bd66fe0af8000000000000000000000000000000000000000000000000): SqliteFailure(Error { code: DatabaseBusy, extended_code: 5 }, Some("database is locked"))
Process abort due to thread panic
   0: stacks_node::main::{{closure}}
             at testnet/stacks-node/src/main.rs:56
   1: std::panicking::rust_panic_with_hook
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:573
   2: std::panicking::begin_panic::{{closure}}
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:498
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/sys_common/backtrace.rs:153
   4: std::panicking::begin_panic
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:497
   5: blockstack_lib::vm::database::sqlite::sqlite_put
             at src/vm/database/sqlite.rs:47
      blockstack_lib::vm::database::sqlite::SqliteConnection::put
             at src/vm/database/sqlite.rs:79
   6: <blockstack_lib::vm::database::marf::MarfedKV as blockstack_lib::vm::database::marf::ClarityBackingStore>::put_all
             at src/vm/database/marf.rs:522
   7: blockstack_lib::vm::database::key_value_wrapper::RollbackWrapper::commit
             at src/vm/database/key_value_wrapper.rs:246
   8: blockstack_lib::vm::clarity::ClarityTransactionConnection::commit
             at src/vm/clarity.rs:851
   9: blockstack_lib::chainstate::stacks::db::transactions::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_transaction
             at src/chainstate/stacks/db/transactions.rs:882
  10: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_block_transactions
             at src/chainstate/stacks/db/blocks.rs:3676
      blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::append_block
             at src/chainstate/stacks/db/blocks.rs:3880
      blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_next_staging_block
             at src/chainstate/stacks/db/blocks.rs:4300
  11: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_blocks
             at src/chainstate/stacks/db/blocks.rs:4458
  12: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,N,U>::process_ready_blocks
             at src/chainstate/coordinator/mod.rs:642
  13: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,N,U>::handle_new_stacks_block
             at src/chainstate/coordinator/mod.rs:433
      blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,blockstack_lib::chainstate::coordinator::comm::ArcCounterCoordinatorNotices,blockstack_lib::chainstate::coordinator::OnChainRewardSetProvider>::run
             at src/chainstate/coordinator/mod.rs:303
  14: stacks_node::run_loop::neon::RunLoop::start::{{closure}}
             at testnet/stacks-node/src/run_loop/neon.rs:155
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/sys_common/backtrace.rs:137
  15: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/thread/mod.rs:458
      <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panic.rs:308
      std::panicking::try::do_call
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:373
      std::panicking::try
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:337
      std::panic::catch_unwind
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panic.rs:379
      std::thread::Builder::spawn_unchecked::{{closure}}
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/thread/mod.rs:457
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/core/src/ops/function.rs:227
  16: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/alloc/src/boxed.rs:1042
      <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/alloc/src/boxed.rs:1042
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/sys/unix/thread.rs:87
  17: start_thread
  18: __clone

Dumping core for pid 17976
```

If you think my database would be useful please let me know and I can share it (I have it saved).

**Comment by whoabuddy** [comment]
Confirming the same panic for me when I checked on my node this morning, it ran successfully from 10/27 until today.

```
ERRO [1604264859.505701] [src/vm/database/sqlite.rs:46] [ThreadId(6)] Failed to insert/replace (05fbcba1b7171dcdf2bcb59a3139a3cbb663a04cc2ca538439d0069e8302f9e60000000000000000,0000000000000000000000212da97100000000000000000000000000000000000000000000000000): SqliteFailure(Error { code: DatabaseBusy, extended_code: 5 }, Some("database is locked"))
Process abort due to thread panic
```

I have the log and can provide it, but was not running debug due to the length of the competition vs the file size that a debug log creates.  

`git rev-parse HEAD` shows a4deb7ac83e88352293d6df4573d1d0f5989fa53 as the latest commit, planning to update and run the miner again starting today.

**Comment by diwakergupta** [comment]
One more crash (rev: 6a8a30b9966ecbfa5b9bef1119d32302362329d9)

```
   0: backtrace::backtrace::libunwind::trace
             at /Users/diwaker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.50/src/backtrace/libunwind.rs:95
      backtrace::backtrace::trace_unsynchronized
             at /Users/diwaker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.50/src/backtrace/mod.rs:66
   1: backtrace::backtrace::trace
             at /Users/diwaker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.50/src/backtrace/mod.rs:53
   2: backtrace::capture::Backtrace::create
             at /Users/diwaker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.50/src/capture.rs:164
   3: backtrace::capture::Backtrace::new
             at /Users/diwaker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.50/src/capture.rs:128
   4: stacks_node::main::{{closure}}
             at testnet/stacks-node/src/main.rs:56
   5: std::panicking::rust_panic_with_hook
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:573
   6: std::panicking::begin_panic::{{closure}}
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:498
   7: std::sys_common::backtrace::__rust_end_short_backtrace
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:153
   8: std::panicking::begin_panic
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:497
   9: blockstack_lib::vm::database::sqlite::sqlite_put
             at src/vm/database/sqlite.rs:47
  10: blockstack_lib::vm::database::sqlite::SqliteConnection::put
             at src/vm/database/sqlite.rs:79
  11: <blockstack_lib::vm::database::marf::MarfedKV as blockstack_lib::vm::database::marf::ClarityBackingStore>::put_all
             at src/vm/database/marf.rs:522
  12: blockstack_lib::vm::database::key_value_wrapper::RollbackWrapper::commit
             at src/vm/database/key_value_wrapper.rs:246
  13: blockstack_lib::vm::clarity::ClarityTransactionConnection::commit
             at src/vm/clarity.rs:851
  14: blockstack_lib::chainstate::stacks::db::transactions::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_transaction
             at src/chainstate/stacks/db/transactions.rs:882
  15: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_block_transactions
             at src/chainstate/stacks/db/blocks.rs:3676
  16: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::append_block
             at src/chainstate/stacks/db/blocks.rs:3880
  17: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_next_staging_block
             at src/chainstate/stacks/db/blocks.rs:4300
  18: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_blocks
             at src/chainstate/stacks/db/blocks.rs:4458
  19: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,N,U>::process_ready_blocks
             at src/chainstate/coordinator/mod.rs:642
  20: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,N,U>::handle_new_stacks_block
             at src/chainstate/coordinator/mod.rs:433
  21: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,blockstack_lib::chainstate::coordinator::comm::ArcCounterCoordinatorNotices,blockstack_lib::chainstate::coordinator::OnChainRewardSetProvider>::run
             at src/chainstate/coordinator/mod.rs:303
  22: stacks_node::run_loop::neon::RunLoop::start::{{closure}}
             at testnet/stacks-node/src/run_loop/neon.rs:155
  23: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:137
  24: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/mod.rs:458
  25: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panic.rs:308
  26: std::panicking::try::do_call
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:373
  27: ___rust_try
  28: std::panicking::try
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:337
  29: std::panic::catch_unwind
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panic.rs:379
  30: std::thread::Builder::spawn_unchecked::{{closure}}
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/mod.rs:457
  31: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/function.rs:227
  32: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/alloc/src/boxed.rs:1042
      <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/alloc/src/boxed.rs:1042
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/sys/unix/thread.rs:87
  33: __pthread_start
```

Full back trace across all threads attached (via core dump on Mac).
[stacks-node_2020-11-02-092524_dune.txt](https://github.com/blockstack/stacks-blockchain/files/5475751/stacks-node_2020-11-02-092524_dune.txt)

**Comment by jcnelson** [comment]
We should try and use WAL mode and see if this goes away.  Also, we need a unit test that can reliably trigger this crash without WAL mode.

**Comment by jcnelson** [comment]
To fix this particular issue:
* We should integrate the side-store with the MARF in the clarity VM (this closes a race condition as well)
* We should rely on the MARF's unroll logic to deal with block roll-backs, instead of using save-points

**Comment by diwakergupta** [comment]
Another crash from a Xenon miner:

```
ERRO [1606483895.465992] [src/vm/database/sqlite.rs:46] [ThreadId(3)] Failed to insert/replace (8cd4e01e214faa7a5f315d90eb06789a17fdfe361bd76[14/90647]
37f1248720000000000000000,0000000000000000000000385852f400000000000000000000000000000000000000000000000000): SqliteFailure(Error { code: DatabaseBusy,
extended_code: 5 }, Some("database is locked"))
Process abort due to thread panic
INFO [1606483898.490895] [testnet/stacks-node/src/neon_node.rs:800] [ThreadId(1)] Sleeping 5000 before issuing tenure
   0: backtrace::backtrace::libunwind::trace
             at /Users/diwaker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.50/src/backtrace/libunwind.rs:95
      backtrace::backtrace::trace_unsynchronized
             at /Users/diwaker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.50/src/backtrace/mod.rs:66
   1: backtrace::backtrace::trace
             at /Users/diwaker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.50/src/backtrace/mod.rs:53
   2: backtrace::capture::Backtrace::create
             at /Users/diwaker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.50/src/capture.rs:164
   3: backtrace::capture::Backtrace::new
             at /Users/diwaker/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.50/src/capture.rs:128
   4: stacks_node::main::{{closure}}
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/testnet/stacks-node/src/main.rs:56
   5: std::panicking::rust_panic_with_hook
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:573
   6: std::panicking::begin_panic::{{closure}}
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:498
   7: std::sys_common::backtrace::__rust_end_short_backtrace
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:153
   8: std::panicking::begin_panic
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:497
   9: blockstack_lib::vm::database::sqlite::sqlite_put
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/vm/database/sqlite.rs:47
  10: blockstack_lib::vm::database::sqlite::SqliteConnection::put
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/vm/database/sqlite.rs:79
  11: <blockstack_lib::vm::database::marf::MarfedKV as blockstack_lib::vm::database::marf::ClarityBackingStore>::put_all
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/vm/database/marf.rs:522
  12: blockstack_lib::vm::database::key_value_wrapper::RollbackWrapper::commit
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/vm/database/key_value_wrapper.rs:249
  13: blockstack_lib::vm::clarity::ClarityTransactionConnection::commit
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/vm/clarity.rs:851
  14: blockstack_lib::chainstate::stacks::db::transactions::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_transaction
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/chainstate/stacks/db/transactions.rs:882
  15: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_block_transactions
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/chainstate/stacks/db/blocks.rs:3752
  16: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::append_block
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/chainstate/stacks/db/blocks.rs:3978
  17: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_next_staging_block
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/chainstate/stacks/db/blocks.rs:4398
  18: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_blocks
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/chainstate/stacks/db/blocks.rs:4556
  19: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,N,U>::process_ready_blocks
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/chainstate/coordinator/mod.rs:674
  20: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,N,U>::handle_new_stacks_block
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/chainstate/coordinator/mod.rs:459
  21: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,blockstack_lib::chainstate::coordinator::comm::ArcCounterCoordinatorNotices,blocksta
ck_lib::chainstate::coordinator::OnChainRewardSetProvider>::run
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/src/chainstate/coordinator/mod.rs:280
  22: stacks_node::run_loop::neon::RunLoop::start::{{closure}}
             at /Users/diwaker/Vault/blockstack/stacks-blockchain/testnet/stacks-node/src/run_loop/neon.rs:208
  23: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:137
  24: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/mod.rs:458
  25: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panic.rs:308
  26: std::panicking::try::do_call
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:373
  27: ___rust_try
  28: std::panicking::try
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:337
  29: std::panic::catch_unwind
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/panic.rs:379
  30: std::thread::Builder::spawn_unchecked::{{closure}}
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/mod.rs:457
  31: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /Users/diwaker/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/function.rs:227
  32: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/alloc/src/boxed.rs:1042
      <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/alloc/src/boxed.rs:1042
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/sys/unix/thread.rs:87
  33: __pthread_start

Dumping core for pid 89169
```

**Comment by friedger** [comment]
for commit fe81b98d81df289686140941a6d65ac49f2c63f4
```
INFO [1608298105.985967] [src/burnchains/db.rs:338] [ThreadId(86)] Storing new burnchain block, burn_header_hash: 000000000000003e8d3268f9d2705338312a86e113895d42c933fb56a0256537
ERRO [1608298106.017001] [src/vm/database/sqlite.rs:46] [ThreadId(2)] Failed to insert/replace (3a6165fa600155c9ccc797b36e2b05cfcc944945cbf058291bdf0f3a13289f640000000000000000,00000000000000000000000000000000000000000000000000000000000000000000000000000000): SqliteFailure(Error { code: DatabaseBusy, extended_code: 5 }, Some("database is locked"))
Process abort due to thread panic
   0: stacks_node::main::{{closure}}
             at testnet/stacks-node/src/main.rs:57
   1: std::panicking::rust_panic_with_hook
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:573
   2: std::panicking::begin_panic::{{closure}}
             at /home/friedger/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:498
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /home/friedger/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:153
   4: std::panicking::begin_panic
             at /home/friedger/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:497
   5: blockstack_lib::vm::database::sqlite::sqlite_put
             at src/vm/database/sqlite.rs:47
   6: blockstack_lib::vm::database::sqlite::SqliteConnection::put
             at src/vm/database/sqlite.rs:79
   7: <blockstack_lib::vm::database::marf::MarfedKV as blockstack_lib::vm::database::marf::ClarityBackingStore>::put_all
             at src/vm/database/marf.rs:585
   8: blockstack_lib::vm::database::key_value_wrapper::RollbackWrapper::commit
             at src/vm/database/key_value_wrapper.rs:249
   9: blockstack_lib::vm::clarity::ClarityTransactionConnection::commit
             at src/vm/clarity.rs:1000
  10: blockstack_lib::chainstate::stacks::db::transactions::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_transaction
             at src/chainstate/stacks/db/transactions.rs:1116
  11: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_block_transactions
             at src/chainstate/stacks/db/blocks.rs:3875
  12: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::append_block
             at src/chainstate/stacks/db/blocks.rs:4265
  13: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_next_staging_block
             at src/chainstate/stacks/db/blocks.rs:4748
  14: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_blocks
             at src/chainstate/stacks/db/blocks.rs:4910
  15: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,N,U>::process_ready_blocks
             at src/chainstate/coordinator/mod.rs:743
  16: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,N,U>::handle_new_burnchain_block
             at src/chainstate/coordinator/mod.rs:577
  17: blockstack_lib::chainstate::coordinator::ChainsCoordinator<T,blockstack_lib::chainstate::coordinator::comm::ArcCounterCoordinatorNotices,blockstack_lib::chainstate::coordinator::OnChainRewardSetProvider>::run
             at src/chainstate/coordinator/mod.rs:309
  18: stacks_node::run_loop::neon::RunLoop::start::{{closure}}
             at testnet/stacks-node/src/run_loop/neon.rs:216
  19: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /home/friedger/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:137
  20: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /home/friedger/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:458
  21: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /home/friedger/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:308
  22: std::panicking::try::do_call
             at /home/friedger/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:373
  23: __rust_try
  24: std::panicking::try
             at /home/friedger/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:337
  25: std::panic::catch_unwind
             at /home/friedger/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:379
  26: std::thread::Builder::spawn_unchecked::{{closure}}
             at /home/friedger/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:457
  27: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /home/friedger/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227
  28: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/alloc/src/boxed.rs:1042
      <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/alloc/src/boxed.rs:1042
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/sys/unix/thread.rs:87
  29: start_thread
  30: clone

```

**Comment by jcnelson** [comment]
Will be fixed by #2199

**Comment by diwakergupta** [comment]
Fixed by #2199
